### PR TITLE
Docs update and warning frozen defined both

### LINKED ISSUE #9850: Dataclass ConfigDict frozen doesn't work as expected
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

There are two apparent ways to make a dataclass frozen. Coming from using ordinary class-based Pydantic models to dataclass-based ones, I found myself providing ConfigDict(frozen=True) to the dataclass decorator. However, this doesn't work with dataclasses, instead you have to provide this argument to the decorator.

This could be made clearer in the docs, one of the arguments could be removed (pref the ConfigDict one), or the dataclass could default to either of which is set (would be a problem if both were set to opposite values).

Cheers!

### Example Code

```Python
from pydantic import ConfigDict
from pydantic.dataclasses import dataclass

@dataclass(config=ConfigDict(frozen=True))
class Location:
    latitude: float
    longitude: float

location = Location(latitude=0, longitude=0)
hash(location)
# TypeError: unhashable type: 'Location'
Works:

from pydantic.dataclasses import dataclass

@dataclass(frozen=True)
class Location:
    latitude: float
    longitude: float

location = Location(latitude=0, longitude=0)
hash(location)
# -8458139203682520985
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.4
        pydantic-core version: 2.18.4
          pydantic-core build: profile=release pgo=true
                 install path: /Users/gregbrown/Development/carta/app/pipeline_service/.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.3 (main, Jun 11 2024, 09:22:06) [Clang 15.0.0 (clang-1500.3.9.4)]
                     platform: macOS-14.5-arm64-arm-64bit
             related packages: mypy-1.10.0 pydantic-settings-2.3.4 typing_extensions-4.12.2
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by mochi22**
## Change Summary
This PR addresses an issue where defining `frozen` both in the `@dataclass` decorator and `config=ConfigDict(frozen=True)` led to confusing and inconsistent behavior. 
The changes made in this PR aim to clarify the expected behavior and provide better guidance to users.

- docs change
```
    It's not recommended to define both `frozen` in the `dataclass` decorator and `ConfigDict`.
    If both are defined, the setting from the dataclass decorator will take priority.
```

In the dataclasses.md file, there was no specific explanation about frozen, so I added an explanation about the frozen setting in ConfigDict(config.py).

- code change
```
            # It's not recommended to define both, as the setting from the dataclass decorator will take priority.
            if config_wrapper.frozen:
                warn(
                    "Both 'dataclass' decorator and 'config.frozen' are defined. 'config.frozen' take priority.",
                    UserWarning,
                )
```

## Related issue number

fix #9850

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

please review: @sydney-runkle

### CODE REVIEW

**Comment by mochi22** [comment]
Hi @sydney-runkle,
Could you please check the PR when it'll be possible!

File: pydantic/config.py [hunk 1]
```diff
@@ -133,6 +133,8 @@ class User(BaseModel):
     Whether models are faux-immutable, i.e. whether `__setattr__` is allowed, and also generates
     a `__hash__()` method for the model. This makes instances of the model potentially hashable if all the
     attributes are hashable. Defaults to `False`.
+    It's not recommended to define both `frozen` in the `dataclass` decorator and `ConfigDict`.
+    If both are defined, the setting from the dataclass decorator will take priority.
```
Comment (Viicos) [inline pydantic/config.py:137]: I would move this in the `dataclass` decorator docstring, `ConfigDict` shouldn't know about Pydantic dataclasses imo

Comment (Viicos) [reply pydantic/config.py:137]: The current `dataclass` docstring is:

```
frozen: Determines if the generated class should be a 'frozen' `dataclass`, which does not allow its
    attributes to be modified after it has been initialized.
```

which I would rephrase as:

```
frozen: Determines if the generated class should be a 'frozen' `dataclass`, which does not allow its
    attributes to be modified after it has been initialized. If not set, the value from the provided `config` argument will be used (and will default to `False` otherwise).
```

File: pydantic/dataclasses.py [hunk 2]
```diff
@@ -230,6 +231,12 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
         if frozen is not None:
             frozen_ = frozen
         else:
+            # It's not recommended to define both, as the setting from the dataclass decorator will take priority.
+            if config_wrapper.frozen:
+                warn(
+                    "Both 'dataclass' decorator and 'config.frozen' are defined. 'config.frozen' take priority.",
+                    UserWarning,
+                )
```
Comment (Viicos) [inline pydantic/dataclasses.py:239]: This is not the case with your current implementation. In the `else` branch, `frozen` is guaranteed to be `None` (i.e. not defined), so they can't both be defined.

Imo we shouldn't raise any warning here, as long as we document the behavior.

**Review by Viicos — CHANGES_REQUESTED** [review]

File: pydantic/dataclasses.py (hunk 2 above)
Comment (sydney-runkle) [reply pydantic/dataclasses.py:239]: I'm inclined to raise a warning here - makes it more likely that folks can figure out the issue on their own + don't open an issue with a question.

Agreed with @Viicos though, this isn't logically in the right place given the warning.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looks like a good start. We've left a few requests for changes. Thanks for your contribution!

File: pydantic/dataclasses.py (hunk 2 above)
Comment (Viicos) [reply pydantic/dataclasses.py:239]: To be tested, but let's use `stacklevel=3` here, so that the warning is emitted at the right place (it should emit at the decorator usage)

Comment (mochi22) [reply pydantic/dataclasses.py:239]: > Agreed with @Viicos though, this isn't logically in the right place given the warning.

I understand that the warning is displayed if config_wrapper.frozen is True, regardless of the value of frozen.
Is my understanding correct?

Comment (sydney-runkle) [reply pydantic/dataclasses.py:239]: We should only display the warning if `frozen` is set in the dataclass constructor, and it's set in `config`.

Comment (mochi22) [reply pydantic/dataclasses.py]: Thank you! I fixed it!

File: pydantic/dataclasses.py [hunk 3]
```diff
@@ -232,6 +233,13 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
         else:
             frozen_ = config_wrapper.frozen or False
 
+        # It's not recommended to define both, as the setting from the dataclass decorator will take priority.
+        if frozen is not None and config_wrapper.frozen == True:
+            warn(
+                "Both 'dataclass' decorator and 'config.frozen' are defined. 'config.frozen' take priority.",
+                UserWarning,
+            )
```
Comment (Viicos) [inline pydantic/dataclasses.py:241]:
```suggestion
if frozen is not None:
    frozen_ = frozen
    if config_wrapper.frozen:
        # It's not recommended to define both, as the setting from the dataclass decorator will take priority.
        warn(...)
else:
    frozen_ = config_wrapper.frozen or False
```

Maybe?

**Review by Viicos — DISMISSED** [review]
I believe a small test (with `pytest.warns`) would be great. We also need to specify the correct `stacklevel` argument, as per [my comment](https://github.com/pydantic/pydantic/pull/10082#discussion_r1709297098).

File: pydantic/dataclasses.py (hunk 3 above)
Comment (mochi22) [reply pydantic/dataclasses.py]: Thank you for your comments, and I agree with you! 
I have fixed it

File: pydantic/dataclasses.py [hunk 4]
```diff
@@ -229,6 +230,12 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
         # Respect frozen setting from dataclass constructor and fallback to config setting if not provided
         if frozen is not None:
             frozen_ = frozen
+            if config_wrapper.frozen:
+                # It's not recommended to define both, as the setting from the dataclass decorator will take priority.
+                warn(
+                    "Both 'dataclass' decorator and 'config.frozen' are defined. 'config.frozen' take priority.",
+                    UserWarning,
+                )
```
Comment (sydney-runkle) [inline pydantic/dataclasses.py:238]:
```suggestion
                warn(
                    f"`frozen` is set via both the 'dataclass' decorator and `config` for dataclass {cls.__name__}."
                    "This is not recommended. The `frozen` specification on `dataclass` will take priority.",
                    UserWarning,
                )
```

File: pydantic/dataclasses.py [hunk 5]
```diff
@@ -124,7 +125,7 @@ def dataclass(
         order: Determines if comparison magic methods should be generated, such as `__lt__`, but not `__eq__`.
         unsafe_hash: Determines if a `__hash__` method should be included in the class, as in `dataclasses.dataclass`.
         frozen: Determines if the generated class should be a 'frozen' `dataclass`, which does not allow its
-            attributes to be modified after it has been initialized.
+            attributes to be modified after it has been initialized. If not set, the value from the provided `config` argument will be used (and will default to `False` otherwise).
```
Comment (sydney-runkle) [inline pydantic/dataclasses.py:128]: It's the other way around - the `dataclass` argument takes priority...

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Requested 2 other changes. Looks good otherwise, thanks!

File: pydantic/dataclasses.py (hunk 5 above)
Comment (mochi22) [reply pydantic/dataclasses.py:128]: It seems the points you mentioned have already been added

Comment (sydney-runkle) [reply pydantic/dataclasses.py:128]: Not sure what you mean - this comment is still incorrect...

**Comment by sydney-runkle** [comment]
As @Viicos mentioned, I'd love to see a test (and the `stacklevel` specification)

File: pydantic/dataclasses.py (hunk 5 above)
Comment (sydney-runkle) [reply pydantic/dataclasses.py:128]: Ah I see - this is fine.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
@mochi22,

Looks great - can merge once we add a test. Thanks!

**Comment by sydney-runkle** [comment]
I'll go ahead and merge this and add a test in a separate PR

**Review by sydney-runkle — APPROVED** [review]
